### PR TITLE
Upgrade Azure storage blob and Azure file  share

### PR DESCRIPTION
**Pull request by ephraimbuddy**
---
This PR upgrades `azure-storage-blob`  and `azure-storage-file-share` to the latest versions(v12).
These versions are not backward compatible with the current version(v2) that airflow uses, so there are lots of changes that are not backward compatible which I have included in the UPDATING.md.

It turns out snowflake-connector-python is causing boto3 conflict which I later saw in their [issues](https://github.com/snowflakedb/snowflake-connector-python/issues/290) and also still using `azure-storage-blob` [version 2](https://github.com/snowflakedb/snowflake-connector-python/blob/master/tested_requirements/requirements_36.txt).

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: tests/providers/microsoft/azure/hooks/test_azure_storage_blob.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            return_value.upload_blob.assert_called_once_with(
+                b'mydata',
+                'BlockBlob',
+                length=4
+            )
+
+    @mock.patch("airflow.providers.microsoft.azure.hooks.azure_storage_blob.BlobServiceClient")
+    def test_download(self, mock_service):
+        hook = AzureStorageBlobHook(azure_blob_conn_id=self.shared_key_conn_id)
+        container_client = mock_service.return_value.get_container_client.return_value
```
Comment (kaxil) [inline tests/providers/microsoft/azure/hooks/test_azure_storage_blob.py:164]:
```suggestion
        container_client = mock_service.return_value.create_container.return_value
```

File: tests/providers/microsoft/azure/hooks/test_azure_storage_blob.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        blob_client = container_client.get_blob_client.return_value
```
Comment (kaxil) [inline tests/providers/microsoft/azure/hooks/test_azure_storage_blob.py]: You need to add a return value to both of these

Comment (ephraimbuddy) [reply tests/providers/microsoft/azure/hooks/test_azure_storage_blob.py]: Ok. Thanks.

**Comment by kaxil** [comment]
Ya it is a bummer. @potiuk Can we handle different versions of the package in the current providers model?

**Comment by potiuk** [comment]
Untortunately not (yet). We have been toying with a possibility of adding extra wheel package or adding separate requirements.txt per DAG or even per task but it has not been done yet. I will create a an issue describing proposed solution and (re) raise discussion for that in devlist soon.

**Comment by ldacey** [comment]
Hi @kaxil - what did you mean when you said "bummer"? is switching to v12 not possible or easy with Airflow currently?

There were some upgrades to an ffspec and adlfs library that I use which requires the new v12 SDK instead of 2.1. I am afraid to update in case I break stuff with Airflow - I currently have remote logging setup on Azure Blob and that is the destination for most of my data.

**Comment by ephraimbuddy** [comment]
> Hi @kaxil - what did you mean when you said "bummer"? is switching to v12 not possible or easy with Airflow currently? @ephraimbuddy 
> 
> There were some upgrades to an ffspec and adlfs library that I use which requires the new v12 SDK instead of 2.1. I am afraid to update in case I break stuff with Airflow - I currently have remote logging setup on Azure Blob and that is the destination for most of my data.

It was not possible to upgrade to v12 due to version conflicts between some packages

**Comment by ldacey** [comment]
Ahh - is this still the case? I was about to give it a shot today.

Some libraries now depend on the new Azure Blob SDK (fsspec, for example) which might cause issues. I know pandas now has an fsspec implementation for their IO and pyarrow uses it as well to read/write parquet files.

**Pull request closed without merging**